pkg/clustermanager/kube_proxy.go

```diff
@@ -0,0 +1,505 @@
+	iptablesLegacyKubeProxy.ObjectMeta.UID = ""
+	image := kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.ImageRepository + "/kube-proxy" +
+		":" + kcp.Spec.Version
+	iptablesLegacyKubeProxy.Spec.Template.Spec.Containers[0].Image = image
+	iptablesLegacyKubeProxy.Spec.Template.Spec.Affinity = &corev1.Affinity{
+		NodeAffinity: &corev1.NodeAffinity{
+			RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
+				NodeSelectorTerms: []corev1.NodeSelectorTerm{
+					{
+						MatchExpressions: []corev1.NodeSelectorRequirement{
+							{
+								Key:      iptablesLegacyLabel,
+								Operator: corev1.NodeSelectorOpExists,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return iptablesLegacyKubeProxy
+}
+
+func createIPTablesLegacyKubeProxy(ctx context.Context, client client.Client, kcp *controlplanev1.KubeadmControlPlane, originalKubeProxy *appsv1.DaemonSet) error {
+	iptablesLegacyKubeProxy := iptablesLegacyKubeProxyFromCurrentDaemonSet(kcp, originalKubeProxy)
+	if err := client.Create(ctx, iptablesLegacyKubeProxy); err != nil && !apierrors.IsAlreadyExists(err) {
+		return errors.Wrap(err, "creating secondary kube-proxy DS with iptables-legacy for old nodes")
+	}
+
+	return nil
+}
+
+func deleteIPTablesLegacyKubeProxy(ctx context.Context, client client.Client) error {
+	iptablesLegacyKubeProxy := &appsv1.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      iptablesLegacyKubeProxyDSName,
+			Namespace: kubeProxyDSNamespace,
+		},
+	}
+
+	if err := client.Delete(ctx, iptablesLegacyKubeProxy); err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrap(err, "deleting secondary kube-proxy DS with iptables-legacy")
+	}
+
+	return nil
+}
+
+func updateKubeProxyVersion(ctx context.Context, client client.Client, kubeProxy *appsv1.DaemonSet, image string) error {
+	kubeProxy.Spec.Template.Spec.Containers[0].Image = image
+	if err := client.Update(ctx, kubeProxy); err != nil {
+		return errors.Wrap(err, "updating main kube-proxy version before upgrade")
+	}
+
+	return nil
+}
+
+func (u KubeProxyUpgrader) ensureUpdateKubeProxyVersion(ctx context.Context, log logr.Logger, client client.Client, spec *cluster.Spec) error {
+	newKubeProxyImage := spec.VersionsBundle.KubeDistro.KubeProxy.URI
+	kubeProxy, err := getKubeProxy(ctx, client)
+	if err != nil {
+		return err
+	}
+
+	if err := updateKubeProxyVersion(ctx, client, kubeProxy, newKubeProxyImage); err != nil {
+		return err
+	}
+
+	// Wait before first iteration so we don't accept results too early
+	time.Sleep(u.updateKubeProxyBackoff)
+	return retrier.Retry(u.updateKubeProxyRetries, u.updateKubeProxyBackoff, func() error {
+		kubeProxy, err = getKubeProxy(ctx, client)
+		if err != nil {
+			return err
+		}
+
+		currentImage := kubeProxy.Spec.Template.Spec.Containers[0].Image
+		if currentImage == newKubeProxyImage {
+			log.V(4).Info("Kube-proxy image update seems stable", "wantImage", newKubeProxyImage, "currentImage", currentImage)
```

Comment:
How do you ensure stability here? So with these retries is this enough to ensure that the new image has not been reverted? This won't run into race conditions right or can it?

Reply:
It waits for two seconds after every update and then checks if the image is the expected one
This tries to mitigate a race condition where the KCP controller hasn't yet read the skip annotation so it undoes our update to the kube-proxy image. So we repeat this update in a loop and verify that the image doesn't get changed back.

If you are asking if this guarantees correctness: as almost any wait based mechanism, it doesn't. It's still possible, although unlikely, that the KCP controller won't see the skip annotation in 2 seconds and might try to undo the change. The longer the wait, the more confidence, but it will never be 100%. I couldn't figure out any algorithm that guaranteed the KCP controller had read the skip annotation before we update the DS.

Reply:
Is 2 seconds enough to have this check then? Just wondering if we need longer or if KCP modifies another property when it sees the skip annotation. If not, yea then this seems fine.